### ISSUE TITLE
ThreadsRunAllTestsHalfManualTest > shouldRunInMultipleThreads occasionally fails on Travis

### ISSUE THREAD

**Issue by szpak** [mockito/mockito#563]
Started to occur more often on the container-based Travis infrastructure:

```
:extTest:build

org.concurrentmockito.ThreadsRunAllTestsHalfManualTest > shouldRunInMultipleThreads FAILED
    junit.framework.AssertionFailedError: Run in multiple thread failed
        at junit.framework.Assert.fail(Assert.java:57)
        at junit.framework.Assert.assertTrue(Assert.java:22)
        at junit.framework.Assert.assertFalse(Assert.java:39)
        at junit.framework.TestCase.assertFalse(TestCase.java:210)
        at org.concurrentmockito.ThreadsRunAllTestsHalfManualTest.shouldRunInMultipleThreads(ThreadsRunAllTestsHalfManualTest.java:147)

1571 tests completed, 1 failed, 38 skipped

:test FAILED
```

**Comment by bric3** [comment]
Look for additional messages, `org.concurrentmockito.ThreadsRunAllTestsHalfManualTest > shouldRunInMultipleThreads` usually fails when the build is interrupted by travis, in this case there should be a line about the code `137`

**Comment by szpak** [comment]
I had to increase log verbosity and three time I've got:
```
    Thread[226]: error!
    1
    java.lang.AssertionError: 
    Expecting empty but was:<{}>
    	at org.mockito.internal.creation.bytebuddy.CachingMockBytecodeGeneratorTest.validate_simple_code_idea_where_weakhashmap_with_classloader_as_key_get_GCed_when_no_more_references(CachingMockBytecodeGeneratorTest.java:109)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:483)
    	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
    	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    	at org.junit.runners.Suite.runChild(Suite.java:128)
    	at org.junit.runners.Suite.runChild(Suite.java:27)
    	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
    	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
    	at org.junit.runner.JUnitCore.run(JUnitCore.java:105)
    	at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:62)
    	at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:49)
    	at org.concurrentmockito.ThreadsRunAllTestsHalfManualTest$AllTestsRunner.run(ThreadsRunAllTestsHalfManualTest.java:68)
```

The original [output](https://travis-ci.org/mockito/mockito/jobs/152798455#L813-L849) (and [here](https://travis-ci.org/mockito/mockito/jobs/152801599#L693-L729)) is very messy because many tests log tracktraces on stdout or stderr, but it is another issue.

@raphw This test has long sleep and depends on GC. Maybe it shouldn't be run in parallel?

Btw, what was the idea behind this `shouldRunInMultipleThreads` test? To have Mockito own tests be ready to  run in parallel?

**Comment by bric3** [comment]
Interesting

**Comment by bric3** [comment]
Looking at the other traces I think this is just a symptom of the real issue.

**Comment by bric3** [comment]
Could the version of the JDKs be an issue inside containers ?

```
java version "1.8.0_31"
```

```
java version "1.7.0_75"
```

**Comment by szpak** [comment]
I tried it yesterday, but in context of build time. I will check it again regarding test failure.

**Comment by bric3** [comment]
Note on non-containerized builds :

```
java version "1.7.0_101" <- open JDK
java version "1.7.0_80" <- oracle JDK
java version "1.8.0_101"
```

So even travis doc is outdated, OpenJDK is more up-to-date than OracleJDK on non containerised builds. We could revert #566 that I just pushed.

**Comment by TimvdLippe** [comment]
Yes I would like to revert to the openjdk whenever possible.

**Comment by szpak** [comment]
The last 4 builds with Java 1.8.0_101 finished successfully (but it can be also by accident). In the meantime the same test failed twice with Java 6 and 7.

**Comment by bric3** [comment]
And have you tried with recent JDK 7/6 ?

**Comment by szpak** [comment]
There is only Java 7 _75 (2015-01-20) and _80 (2015-04-14) available. I don't know if it is possible to install newer OpenJDK 6 version on Travis.

Btw, the last 3 builds from that branch finished successfully on all Java vesions...

**Comment by paulduffin** [comment]
I've just seen a similar issue on my local machine not consistently though.

```
org.concurrentmockito.ThreadsRunAllTestsHalfManualTest > shouldRunInMultipleThreads FAILED
    java.lang.AssertionError: Run in multiple thread failed for tests expected:<[]> but was:<[class org.mockitousage.verification.BasicVerificationTest]>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at org.concurrentmockito.ThreadsRunAllTestsHalfManualTest.shouldRunInMultipleThreads(ThreadsRunAllTestsHalfManualTest.java:148)
```

**Comment by raphw** [comment]
I siggest to add a property `org.mockito.lock.singleton` that allows to use a single lock for all mock classes as it was done before. This reduces concurrency and can therefore reduce performance but I cannot see how our locking is the root cause for these locks. I assume that there is a tricky scenario with regards to class loading that is not caused by us but triggered via our explicit loading when creating a mock.

As this changes the class loading order, the lock might be a side effect.
